src/algebra/continued_fractions/convergents_equiv.lean

```diff
@@ -0,0 +1,410 @@
+  case nat.zero
+  { suffices : (b * g.h + a) / b = g.h + a / b, by
+      simpa [squash_gcf, s_nth_eq, convergent_eq_conts_a_div_conts_b,
+        (continuants_recurrence_aux s_nth_eq zeroth_continuant_aux_eq_one_zero
+        first_continuant_aux_eq_h_one)],
+    calc
+      (b * g.h + a) / b = b * g.h / b + a / b  : by ring -- requires `field` rather than `division_ring`
+                    ... = g.h + a / b          : by rw (mul_div_cancel_left _ b_ne_zero) },
+  case nat.succ
+  { obtain ⟨⟨pa, pb⟩, s_n'th_eq⟩ : ∃ gp_n', g.s.nth n' = some gp_n', from
+      g.s.ge_stable n'.le_succ s_nth_eq,
+    -- Notations
+    let g' := squash_gcf g (n' + 1),
+    set predConts := g.continuants_aux (n' + 1) with succ_n'th_conts_aux_eq,
+    set ppredConts := g.continuants_aux n' with n'th_conts_aux_eq,
+    let pA := predConts.a, let pB := predConts.b, let ppA := ppredConts.a, let ppB := ppredConts.b,
+    set predConts' := g'.continuants_aux (n' + 1) with succ_n'th_conts_aux_eq',
+    set ppredConts' := g'.continuants_aux n' with n'th_conts_aux_eq',
+    let pA' := predConts'.a, let pB' := predConts'.b, let ppA' := ppredConts'.a,
+    let ppB' := ppredConts'.b,
+    -- first compute the convergent of the squashed gcf
+    have : g'.convergents (n' + 1)
+         = ((pb + a / b) * pA' + pa * ppA') / ((pb + a / b) * pB' + pa * ppB'), by
+    { have : g'.s.nth n' = some ⟨pa, pb + a / b⟩, by
+        simpa only [squash_nth_gcf] using
+          (squash_seq_nth_of_not_terminated s_n'th_eq s_nth_eq),
+      rw [convergent_eq_conts_a_div_conts_b,
+        (continuants_recurrence_aux this n'th_conts_aux_eq'.symm succ_n'th_conts_aux_eq'.symm)], },
+    rw this,
+    -- then compute the convergent of the original gcf by recursively unfolding the continuants
+    -- computation twice
+    have : g.convergents (n' + 2)
+         = (b * (pb * pA + pa * ppA) + a * pA) / (b * (pb * pB + pa * ppB) + a * pB), by
+    {
+      -- use the recurrence once
+      have : g.continuants_aux (n' + 2) = ⟨pb * pA + pa * ppA, pb * pB + pa * ppB⟩, from
+        continuants_aux_recurrence s_n'th_eq n'th_conts_aux_eq.symm succ_n'th_conts_aux_eq.symm,
+      -- and a second time
+      rw [convergent_eq_conts_a_div_conts_b,
+        (continuants_recurrence_aux s_nth_eq succ_n'th_conts_aux_eq.symm this)] },
+    rw this,
+    -- now use the fact that the continuants before position n are equal.
+    suffices : ((pb + a / b) * pA + pa * ppA) / ((pb + a / b) * pB + pa * ppB)
+             = (b * (pb * pA + pa * ppA) + a * pA) / (b * (pb * pB + pa * ppB) + a * pB), by
+    { obtain ⟨eq1, eq2, eq3, eq4⟩ : pA' = pA ∧ pB' = pB ∧ ppA' = ppA ∧ ppB' = ppB, by
+        simp [*, (continuants_aux_eq_continuants_aux_squash_gcf_of_le $ le_refl $ n' + 1).symm,
+          (continuants_aux_eq_continuants_aux_squash_gcf_of_le n'.le_succ).symm],
+      symmetry,
+      simpa only [eq1, eq2, eq3, eq4] },
+    calc
+          ((pb + a / b) * pA + pa * ppA) / ((pb + a / b) * pB + pa * ppB)
```

Comment:
I wonder if the goal can be closed by something like `field_simp [b_ne_zero]; ring`.

Reply:
It does not sadly. I had a shorter proof using `ring` and friends, but that made the proof way slower than it is now, so I would like to keep it this way.

Reply:
I just had a look, and the fact that `field_simp` doesn't work is because there is a cancellation of `b/b` it can't see. It works if one adds it by hand, like in
```lean
        suffices : ((pb + a / b) * pA + pa * ppA) / ((pb + a / b) * pB + pa * ppB)
             = (b * (pb * pA + pa * ppA) + a * pA) / (b * (pb * pB + pa * ppB) + a * pB) * b/b, by
    { obtain ⟨eq1, eq2, eq3, eq4⟩ : pA' = pA ∧ pB' = pB ∧ ppA' = ppA ∧ ppB' = ppB, by
        simp [*, (continuants_aux_eq_continuants_aux_squash_gcf_of_le $ le_refl $ n' + 1).symm,
          (continuants_aux_eq_continuants_aux_squash_gcf_of_le n'.le_succ).symm],
      symmetry,
      simpa only [eq1, eq2, eq3, eq4, mul_div_cancel'' _  b_ne_zero] },
    field_simp [b_ne_zero],
    congr' 1;
    ring } }
```
This is very quick on my computer.

Reply:
I see. Isn't this like a non-terminal simp though and hence should strictly be avoided?

Reply:
Non-terminal simps are bad if they are followed by something which depends on the precise shape of the result (like `rw` or `apply`) but they are fine if they are followed by tactics that do not really depend on the shape like `ring` here. `field_simp, ring` is the standard idiom to automate computations with divisions.

Reply:
Changed it.